lib/Dialect/Moore/Transforms/DeleteLocalVar.cpp

```diff
@@ -0,0 +1,123 @@
+//===- DeleteLocalVar.cpp - Delete local temporary variables --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DeleteLocalVar pass.
+// Use to collect net/variable declarations and bound a value to them.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "llvm/ADT/ScopedHashTable.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace moore;
+
+namespace {
+struct DeleteLocalVarPass : public DeleteLocalVarBase<DeleteLocalVarPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+namespace {
+struct DLV { // Delete Local Variables
+  /// All local temporary variables at the `moore.procedure`, `scf.if`,
+  /// `scf.while` body region/block need to be deleted, they won't be reserved
+  /// at the low level.
+  void deleteLocalVar(DenseSet<Operation *> &localVars,
+                      DenseSet<Operation *> &users, mlir::Region &region);
+
+  /// The `valueSymbols` used to collect nets/variables and their initial or
+  /// assignment values.
+  using ValueSymbols = llvm::ScopedHashTable<Operation *, Value>;
+  using ValueSymbolScope = ValueSymbols::ScopeTy;
+  ValueSymbols valueSymbols;
+};
+} // namespace
+
+void DLV::deleteLocalVar(DenseSet<Operation *> &localVars,
+                         DenseSet<Operation *> &users, mlir::Region &region) {
+  ValueSymbolScope scope(valueSymbols);
+
+  for (auto &op : region.getOps()) {
+    // Assume not to assign a new value to the local variable.
+    bool isNewValue = false;
+
+    TypeSwitch<Operation *, void>(&op)
+        // Collect all local variables and their initial values
+        // if exist and all users.
+        .Case<VariableOp>([&](auto op) {
+          valueSymbols.insert(op, op.getInitial());
+          localVars.insert(op);
+          for (auto *user : op->getUsers())
+            users.insert(user);
+        })
+        // Update the values of local variables.
+        .Case<BlockingAssignOp, NonBlockingAssignOp>([&](auto op) {
+          auto destOp = op.getDst().getDefiningOp();
+          auto srcOp = op.getSrc().getDefiningOp();
+
+          // For example, `a = 1`, `b = a`, `c = b` both are local variable.
+          // After updating: `a = 1`, `b = 1`, `c = 1`.
+          if (valueSymbols.count(destOp)) {
+            if (auto srcValue = valueSymbols.lookup(srcOp))
+              valueSymbols.insert(destOp, srcValue);
+            else
+              valueSymbols.insert(destOp, op.getSrc());
+
+            isNewValue = true;
+          }
```

Comment:
I think all you need is an `else` branch here that emits an error. Something like "cannot assign this expression". This will make sure that the user sees a proper error for anything that the pass does not yet fully support.

Reply:
If I don't misunderstand, I need to add an `else` branch. Like `if(valueSymbols.count(destOp)){...} else{error:...}` . But this will lead to global variables also triggering this error.